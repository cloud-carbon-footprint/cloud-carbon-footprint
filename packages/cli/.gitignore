dist
*.csv
*.cache.json
estimates.cache.*.json
*.snap
.env
.env.production
!aws_input.test.csv
!gcp_input.test.csv
!azure_input.test.csv
!incorrect_input.test.csv
!on_premise_data_input.test.csv