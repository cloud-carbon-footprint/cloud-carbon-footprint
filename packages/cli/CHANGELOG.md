# @cloud-carbon-footprint/cli

## 1.1.0

### Minor Changes

- 61332214: Adds CLI command for creating a lookup table to be used in ETL pipelines
- 52237bc6: Adds additional dashboard for viewing cloud provider recommendations including refactoring filters for reusability

  For updating the create-app templates, please refer to the following [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/d1f9a94ea88e6f9210a781e16df18b9f64a0b03d).

### Patch Changes

- Updated dependencies [f3569daa]
- Updated dependencies [61332214]
- Updated dependencies [52237bc6]
  - @cloud-carbon-footprint/app@0.3.0
  - @cloud-carbon-footprint/common@1.2.0

## 1.0.2

### Patch Changes

- @cloud-carbon-footprint/app@0.2.1

## 1.0.1

### Patch Changes

- 52a8b3c1: Adds support for specifying recommendation target for AWS via API parameter

  Please refer to this [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/404c77796ae838766dc8007e18daee2c7526f6ed) to update create app templates.

- Updated dependencies [f75cf08f]
- Updated dependencies [52a8b3c1]
  - @cloud-carbon-footprint/app@0.2.0
  - @cloud-carbon-footprint/common@1.1.0

## 1.0.0

### Major Changes

- 91ed3d75: Update variables GCP_BILLING_ACCOUNT_ID/GCP_BILLING_ACCOUNT_NAME to be GCP_BILLING_PROJECT_ID/GCP_BILLING_PROJECT_NAME

  This is a major update for packages: api, cli, common. Please refer to this [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/aab00ea5a395d94ba5ce1424ffa33abbafd7ed58) for create-app template updates as they are breaking changes.

### Patch Changes

- bbceed8b: Updates dependencies: @types/fs-extra, typescript, husky and @types/node
- Updated dependencies [13f39ac7]
- Updated dependencies [91ed3d75]
- Updated dependencies [72fc2752]
- Updated dependencies [e76d5fdd]
- Updated dependencies [bbceed8b]
  - @cloud-carbon-footprint/app@0.1.0
  - @cloud-carbon-footprint/common@1.0.0

## 0.4.2

### Patch Changes

- 3b42775b: Updates @types/node to 15.12.4
- Updated dependencies [e93d31ec]
- Updated dependencies [3b42775b]
- Updated dependencies [53366130]
  - @cloud-carbon-footprint/common@0.0.3
  - @cloud-carbon-footprint/app@0.0.4

## 0.4.1

### Patch Changes

- Updated dependencies [8e24c32b]
  - @cloud-carbon-footprint/common@0.0.2
  - @cloud-carbon-footprint/app@0.0.3

## 0.4.0

### Minor Changes

- ababb826: Extracts two new packages app and common to avoid circular dependancies and make it easier to extract cloud provider packages
- e84a4c7a: Extract logic into the new packages: app, common, gcp, aws, azure:

  There are many files that have been updated/extracted.
  In order to update create-app templates, refer to the follow [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/8c6aaed52e9f3949e134852986d50362aad3367a).

  The following changes were made to,
  'packages/create-app/templates/default-app/packages/client/tsconfig.json':

  ```diff
      // ...
          "skipLibCheck": true,
          "esModuleInterop": true,
          "allowSyntheticDefaultImports": true,
  -        "strict": true,
          "forceConsistentCasingInFileNames": true,
          "module": "esnext",
          "moduleResolution": "node",
          "resolveJsonModule": true,
  -        "isolatedModules": true,
          "noEmit": true,
          "jsx": "react-jsx",
  -        "noFallthroughCasesInSwitch": true
  +        "noFallthroughCasesInSwitch": true,
  +        "strict": false,
  +        "isolatedModules": true
        },
  -      "include": [
  -        "src",
  -        "node_modules/apexcharts/types/apexcharts.d.ts"
  -      ]
  +      "include": ["src"]
      }
      // ...
  ```

  Additionally, the following dependencies have been updated and should also be updated in their respective template package.json file:

  - @cloud-carbon-footprint root package.json:
    - "@types/fs-extra": "^9.0.11"
    - "concurrently": "^6.2.0"
    - "marked": ">=2.0.5"
  - @cloud-carbon-footprint/api and @cloud-carbon-footprint/cli:
    - "dotenv": "^10.0.0"
    - "@cloud-carbon-footprint/app": Can be added with `yarn up @cloud-carbon-footprint/app`
    - "@cloud-carbon-footprint/common": Can be added with `yarn up @cloud-carbon-footprint/common`
  - @cloud-carbon-footprint/client:
    - "dotenv": "^10.0.0"
    - "@testing-library/react-hooks": "^7.0.0"
    - "concurrently": "^6.2.0"
    - "@cloud-carbon-footprint/common": Can be added with `yarn up @cloud-carbon-footprint/common`

### Patch Changes

- c7fa7db0: Updates dependencies to the latest
- Updated dependencies [f9fbcb4c]
- Updated dependencies [c7fa7db0]
- Updated dependencies [ababb826]
- Updated dependencies [e84a4c7a]
  - @cloud-carbon-footprint/core@0.8.0

## 0.3.3

### Patch Changes

- b63d8a67: The default `aws-sdk` dependency was bumped to `"^2.890.0"`,
- Updated dependencies [56bb6da6]
- Updated dependencies [68365cbf]
- Updated dependencies [b63d8a67]
- Updated dependencies [8df5703b]
- Updated dependencies [3e2f876d]
- Updated dependencies [3abe3dca]
- Updated dependencies [29f48e7c]
- Updated dependencies [370c509d]
- Updated dependencies [7d523b59]
  - @cloud-carbon-footprint/core@0.7.0

## 0.3.2

### Patch Changes

- 81dfde6c: Improves .env.template files with links to cofiguration glossary
- b0151b7d: Improves GCP guided-install prompts in the CLI and create-app packages
- Updated dependencies [8b4c992d]
- Updated dependencies [216ea2ec]
  - @cloud-carbon-footprint/core@0.6.0

## 0.3.1

### Patch Changes

- 8175b41d: updates packages with readmes and metadata
- Updated dependencies [8175b41d]
- Updated dependencies [0ccce96b]
  - @cloud-carbon-footprint/core@0.5.0

## 0.3.0

### Minor Changes

- 23d841e: Adds support for guided install command in the cli package.

### Patch Changes

- Updated dependencies [6c620db]
- Updated dependencies [f3d4c8a]
- Updated dependencies [c5f28fe]
- Updated dependencies [906f14e]
- Updated dependencies [bb82d7b]
- Updated dependencies [033a504]
- Updated dependencies [9b10f3b]
- Updated dependencies [82aeb1e]
  - @cloud-carbon-footprint/core@0.4.0

## 0.2.1

### Patch Changes

- a0ac965: Fixes failing tests when running them in a different timezone
- Updated dependencies [beabadf]
- Updated dependencies [e2ac070]
- Updated dependencies [29e9897]
- Updated dependencies [75176d7]
- Updated dependencies [ce0f249]
- Updated dependencies [29e9897]
- Updated dependencies [a0ac965]
  - @cloud-carbon-footprint/core@0.3.0

## 0.2.0

### Minor Changes

- 96cb023: Adds support for Microsoft Azure, as well as updates to the microsite.

### Patch Changes

- Updated dependencies [96cb023]
  - @cloud-carbon-footprint/core@0.2.0

## 0.1.0

### Minor Changes

- 3b3bec3: Update packages to have base version
