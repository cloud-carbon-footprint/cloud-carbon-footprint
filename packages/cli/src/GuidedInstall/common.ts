/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import { promisify } from 'util'
import { exec as execCb } from 'child_process'
import dotenv from 'dotenv'
import fs from 'fs-extra'
import { confirm, input, list, prompt } from 'typed-prompts'
import { prop } from 'ramda'

const exec = promisify(execCb)
const stat = promisify(fs.stat)

export const log = (m: string, leading = true): void =>
  console.log(`${leading ? '\n' : ''}# ${m}\n`)

export const lineBreak = (): void => console.log()

export const listPrompt = (
  message: string,
  options: string[],
): Promise<string> =>
  prompt<{ key: string }>([list('key', message, options)]).then(prop('key'))

export const confirmPrompt = (
  message: string,
  question = 'Is this step complete?',
  requireYes = true,
): Promise<boolean> =>
  prompt<{ key: boolean }>([confirm('key', message + `\n${question}`)])
    .then(prop('key'))
    .then((result) => {
      if (requireYes && !result) {
        log('Please try again when you have completed this step.')
        process.exit(0)
      }
      console.log()
      return result
    })

export const inputPrompt = (
  message: string,
  required = true,
): Promise<string> =>
  prompt<{ key: string }>([input('key', message)])
    .then(prop('key'))
    .then((result: string) => {
      if (required && !result) {
        log('Please enter a value.')
        return inputPrompt(message, required)
      }
      return result
    })

export const runCmd = async (cmd: string): void => {
  try {
    await exec(cmd)
  } catch (error) {
    process.stdout.write(error.stderr)
    process.stdout.write(error.stdout)
    throw new Error(`Could not execute command ${cmd}`)
  }
}
// will load existing .env and overwrite the given keys
export const createEnvFile = async (
  dir: string,
  env: { [key: string]: string },
): void => {
  const path = `${dir}.env`
  log(`Creating ${path}...`, false)
  const exists = await stat(path)
    .then(() => true)
    .catch(() => false)

  const existing = exists ? dotenv.config({ path }) : {}
  if (existing.error) {
    throw existing.error
  }

  const newEnv = { ...existing.parsed, ...env }
  await fs.writeFile(
    path,
    Object.entries(newEnv).reduce(
      (config, [k, v]) => `${config}${k}=${v}\n`,
      '',
    ),
  )
  log('...done', false)
}
