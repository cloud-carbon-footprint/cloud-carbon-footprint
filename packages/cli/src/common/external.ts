/*
 * Â© 2021 Thoughtworks, Inc.
 */

/* istanbul ignore file */
import { exec as _exec } from 'child_process'
import fs from 'fs-extra'
import { promisify } from 'util'
import { prompt as _prompt, Question } from 'typed-prompts'
import { resolve as _resolve } from 'path'
import dotenv from 'dotenv'

export const exec = promisify(_exec)

export const stat = promisify(fs.stat)

export const writeFile = fs.writeFile

export const log = (m: string, leading = true): void =>
  console.log(`${leading ? '\n' : ''}# ${m}\n`)

export const lineBreak = (): void => console.log()

export const exit = (code = 0): void => process.exit(code)

export const prompt = <T>(question: Question): Promise<T> =>
  _prompt<T>([question])

export const readConfig = (
  path: string,
): { parsed?: { [key: string]: string }; error?: Error } =>
  dotenv.config({ path })

export const resolve = _resolve

export const runCmd = async (cmd: string): Promise<void> => {
  try {
    await exec(cmd)
  } catch (error) {
    process.stdout.write(error.stderr)
    process.stdout.write(error.stdout)
    throw new Error(`Could not execute command ${cmd}`)
  }
}
