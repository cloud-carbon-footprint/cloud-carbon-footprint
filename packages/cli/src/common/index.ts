/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { confirm, input, list } from 'typed-prompts'
import { prop } from 'ramda'
import {
  exit,
  lineBreak,
  log,
  prompt,
  readConfig,
  resolve,
  stat,
  writeFile,
} from './external'

export type EnvConfig = { [key: string]: string }

export const microsite = 'https://cloud-carbon-footprint.org'

export const listPrompt = (
  message: string,
  options: string[],
): Promise<string> =>
  prompt<{ key: string }>(list('key', message, options)).then(prop('key'))

export const confirmPrompt = (
  message: string,
  question = 'Is this step complete?',
  requireYes = true,
): Promise<boolean> =>
  prompt<{ key: boolean }>(confirm('key', message + `\n${question}`))
    .then(prop('key'))
    .then((result) => {
      if (requireYes && !result) {
        log('Please try again when you have completed this step.')
        exit()
      }
      lineBreak()
      return result
    })

export const inputPrompt = (
  message: string,
  required = true,
): Promise<string> =>
  prompt<{ key: string }>(input('key', message))
    .then(prop('key'))
    .then((result: string) => {
      if (required && !result) {
        log('Please enter a value.')
        return inputPrompt(message, required)
      }
      return result
    })

export const createEnvFile = async (
  dir: string,
  env: { [key: string]: string },
): Promise<void> => {
  const path = resolve(`${dir}.env`)
  log(`Creating/updating ${path}...`, false)
  const exists = await stat(path)
    .then(() => true)
    .catch(() => false)

  const existing = exists ? readConfig(path) : { parsed: undefined }
  if (existing.error) {
    throw existing.error
  }

  const newEnv = { ...existing.parsed, ...env }
  await writeFile(
    path,
    Object.entries(newEnv).reduce(
      (config, [k, v]) => `${config}${k}=${v}\n`,
      '',
    ),
  )
  log('...done', false)
}
