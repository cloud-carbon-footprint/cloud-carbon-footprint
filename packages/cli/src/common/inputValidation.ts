/*
 * Â© 2021 Thoughtworks, Inc.
 */
import {
  EstimationRequestValidationError,
  OnPremiseDataInput,
} from '@cloud-carbon-footprint/common'

type cliDataInput = {
  groupBy: string
}

export const validateOnPremiseInput = (inputData: OnPremiseDataInput[]) => {
  inputData.map((inputRow: OnPremiseDataInput) => {
    if (
      !inputRow.cpuDescription ||
      !inputRow.machineType ||
      !inputRow.memory ||
      !inputRow.startTime ||
      !inputRow.endTime ||
      !inputRow.dailyUptime ||
      !inputRow.weeklyUptime ||
      !inputRow.monthlyUptime ||
      !inputRow.annualUptime
    ) {
      throw new Error(
        'Input data is incorrect. Please check your input data file and try again.',
      )
    }
  })
}

export const validateCliInput = (inputData: cliDataInput) => {
  const groupingOptions = ['day', 'dayAndService', 'service']
  if (!groupingOptions.includes(inputData.groupBy)) {
    throw new EstimationRequestValidationError(
      `GroupBy param is incorrect. Please specify one of the following grouping methods: ${groupingOptions.join(
        ' | ',
      )}`,
    )
  }
}
