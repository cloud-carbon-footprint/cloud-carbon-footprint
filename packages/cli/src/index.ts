/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import cli from './cli'
import * as console from 'console'

cli().then(console.log)
