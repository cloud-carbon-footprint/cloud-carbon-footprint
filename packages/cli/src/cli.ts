/*
 * Â© 2021 Thoughtworks, Inc.
 */

import commander from 'commander'
import moment from 'moment'
import path from 'path'
import * as process from 'process'

import {
  App,
  createValidFootprintRequest,
  MongoDbCacheManager,
} from '@cloud-carbon-footprint/app'
import { EstimationResult, configLoader } from '@cloud-carbon-footprint/common'

import EmissionsByDayAndServiceTable from './EmissionsByDayAndServiceTable'
import EmissionsByServiceTable from './EmissionsByServiceTable'
import EmissionsByDayTable from './EmissionsByDayTable'
import CliPrompts from './CliPrompts'
import { exportToCSV } from './CSV'
import { validateCliInput } from './common/inputValidation'

export default async function cli(argv: string[] = process.argv) {
  const program = new commander.Command()
  program.storeOptionsAsProperties(false)

  program
    .option('-s, --startDate <string>', 'Start date in ISO format')
    .option('-e, --endDate <string>', 'End date in ISO format')
    .option(
      '-g, --groupBy <string>',
      'Group results by day or service. Default is day.',
    )
    .option('-i, --interactive', 'Use interactive CLI prompts')
    .option(
      '-f, --format <string>',
      'How to format the results [table, csv]. Default is table.',
    )

  program.parse(argv)

  let startDate, endDate
  let groupBy: string
  let format: string

  if (program.opts().interactive) {
    ;[startDate, endDate, groupBy, format] = await CliPrompts()
  } else {
    const programOptions = program.opts()
    startDate = programOptions.startDate
    endDate = programOptions.endDate
    groupBy = programOptions.groupBy
    format = programOptions.format

    if (!groupBy) {
      console.warn(
        'GroupBy parameter not specified, adopting "day" as the default.',
      )
      groupBy = 'day'
    }
  }
  validateCliInput({ groupBy })

  const estimationRequest = createValidFootprintRequest({
    startDate,
    endDate,
    groupBy: 'day', // So that estimates are cached the same regardless of table grouping method
  })

  if (configLoader().CACHE_MODE === 'MONGODB') {
    await MongoDbCacheManager.createDbConnection()
  }

  const { table, colWidths } = await new App()
    .getCostAndEstimates(estimationRequest)
    .then((estimations: EstimationResult[]) => {
      if (groupBy === 'service') {
        return EmissionsByServiceTable(estimations)
      }
      if (groupBy === 'day') {
        return EmissionsByDayTable(estimations)
      }
      return EmissionsByDayAndServiceTable(estimations)
    })

  if (configLoader().CACHE_MODE === 'MONGODB') {
    await MongoDbCacheManager.mongoClient.close()
    console.log('MongoDB connection closed')
  }

  if (format === 'csv') {
    const filePath = path.join(
      process.cwd(),
      `results-${moment().utc().format('YYYY-MM-DD-HH:mm:ss')}.csv`,
    )
    exportToCSV(table, filePath)
    return `File saved to: ${filePath}`
  } else {
    return table
      .map((row: string[]) =>
        row.reduce(
          (acc, data, col) => acc + `| ${data}`.padEnd(colWidths[col]),
          '',
        ),
      )
      .join('\n')
  }
}
