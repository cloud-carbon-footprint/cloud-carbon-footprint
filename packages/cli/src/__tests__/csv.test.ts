/*
 * Â© 2021 Thoughtworks, Inc.
 */

import path from 'path'
import fs from 'fs'
import AWSMock from 'aws-sdk-mock'
import AWS from 'aws-sdk'

import {
  mockAwsCloudWatchGetMetricData,
  mockAwsCostExplorerGetCostAndUsage,
} from './fixtures/awsMockFunctions'
import cli from '../cli'
import { EstimationResult } from '@cloud-carbon-footprint/common'

const mockGetCostAndEstimates = jest.fn()
const mockGetFilterData = jest.fn()
const mockGetEmissionsFactors = jest.fn()

jest.mock('@cloud-carbon-footprint/app', () => ({
  ...(jest.requireActual('@cloud-carbon-footprint/app') as Record<
    string,
    unknown
  >),
  App: jest.fn().mockImplementation(() => {
    return {
      getCostAndEstimates: mockGetCostAndEstimates,
      getEmissionsFactors: mockGetEmissionsFactors,
      getFilterData: mockGetFilterData,
    }
  }),
  cache: jest.fn(),
}))

jest.mock('@cloud-carbon-footprint/common', () => ({
  ...(jest.requireActual('@cloud-carbon-footprint/common') as Record<
    string,
    unknown
  >),
  Logger: jest.fn().mockReturnValue({
    debug: jest.fn(),
    info: jest.fn(),
    error: jest.fn(),
    warn: jest.fn(),
  }),
  configLoader: jest.fn().mockImplementation(() => {
    return {
      AWS: {
        accounts: [{ id: '12345678', name: 'test AWS account' }],
        NAME: 'AWS',
        CURRENT_REGIONS: ['us-east-1', 'us-east-2'],
        CURRENT_SERVICES: [
          {
            key: 'ebs',
            name: 'EBS',
          },
          {
            key: 's3',
            name: 'S3',
          },
          {
            key: 'ec2',
            name: 'EC2',
          },
          {
            key: 'elasticache',
            name: 'ElastiCache',
          },
          {
            key: 'rds',
            name: 'RDS',
          },
          {
            key: 'lambda',
            name: 'Lambda',
          },
        ],
        authentication: {
          mode: 'GCP',
          options: {
            targetRoleName: 'test-target',
            proxyAccountId: 'test-account-id',
            proxyRoleName: 'test-role-name',
          },
        },
      },
      GCP: {
        projects: [
          { id: '987654321', name: 'test GCP account' },
          { id: '987654321', name: 'test GCP account 2' },
        ],
        NAME: 'GCP',
        CURRENT_SERVICES: [{ key: 'testService', name: 'service' }],
        CURRENT_REGIONS: ['us-east1', 'us-west1', 'us-central1'],
        CACHE_BUCKET_NAME: 'test-bucket-name',
      },
    }
  }),
}))

beforeAll(() => {
  AWSMock.setSDKInstance(AWS)
})

afterEach(() => {
  AWSMock.restore()
})

describe('csv test', () => {
  const start = '2020-07-01'
  const end = '2020-07-07'
  const rawRequest = [
    'executable',
    'file',
    '--startDate',
    start,
    '--endDate',
    end,
  ]

  let outputFilePath: string

  beforeEach(() => {
    jest.spyOn(Date, 'now').mockImplementation(() => 1596660091000)
    outputFilePath = path.join(process.cwd(), 'results-2020-08-05-20:41:31.csv')
    const expectedResponse: EstimationResult[] = []
    mockGetCostAndEstimates.mockResolvedValueOnce(expectedResponse)
  })

  afterEach(() => {
    try {
      fs.unlinkSync(outputFilePath)
    } catch (err) {
      console.error(err)
    }
    jest.restoreAllMocks()
  })

  test('formats table into csv file', async () => {
    mockAwsCloudWatchGetMetricData()
    mockAwsCostExplorerGetCostAndUsage()

    await cli([...rawRequest, '--format', 'csv', '--groupBy', 'dayAndService'])

    expect(fs.existsSync(outputFilePath)).toBe(true)
    expect(fs.readFileSync(outputFilePath).toString()).toMatchSnapshot()
  })
})
