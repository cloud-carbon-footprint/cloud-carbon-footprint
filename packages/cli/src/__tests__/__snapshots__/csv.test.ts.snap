// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csv test formats table into csv file 1`] = `
"Date (UTC),EBS kilowatt hours,EBS metric tons CO2e Emissions,S3 kilowatt hours,S3 metric tons CO2e Emissions,EC2 kilowatt hours,EC2 metric tons CO2e Emissions,ElastiCache kilowatt hours,ElastiCache metric tons CO2e Emissions,RDS kilowatt hours,RDS metric tons CO2e Emissions,SUM kilowatt hours,SUM metric tons CO2e Emissions
2020-07-01,1.01,0.000461,0.04,0.000019,19.74,0.008971,2.56,0.001164,4.03,0.001833,27.39,0.012448
2020-07-02,2.03,0.000921,0.00,0.000000,0.00,0.000000,9.47,0.004302,228.19,0.103715,239.69,0.108938
Total,3.04,0.001382,0.04,0.000019,19.74,0.008971,12.03,0.005467,232.23,0.105548,267.08,0.121386
"
`;
